src/Microsoft.VisualStudio.Services.Agent/Blob/BlobStoreUtils.cs

```diff
@@ -26,6 +22,151 @@ namespace Microsoft.VisualStudio.Services.Agent.Blob
+            var uploadSession = dedupClient.CreateUploadSession(keepUntilRef, tracer, FileSystem.Instance);
+
+            using (var reportingCancelSrc = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken))
+            {
+                // Log stats
+                Task reportingTask = null;
+                if (enableReporting)
+                {
+                    reportingTask = StartReportingTask(traceOutput, (long)rootNode.TransitiveContentBytes, uploadSession, reportingCancelSrc);
+                }
+
+                // Upload the chunks
+                var uploadRecord = clientTelemetry.CreateRecord<BlobStoreTelemetryRecord>(telemetryRecordFactory);
+                await clientTelemetry.MeasureActionAsync(
+                    record: uploadRecord,
+                    actionAsync: async () => await AsyncHttpRetryHelper.InvokeAsync(
+                            async () =>
+                            {
+                                await uploadSession.UploadAsync(rootNode, fileDedupIds, cancellationToken);
+                                return uploadSession.UploadStatistics;
+                            },
+                            maxRetries: 3,
+                            tracer: tracer,
+                            canRetryDelegate: e => true, // this isn't great, but failing on upload stinks, so just try a couple of times
+                            cancellationToken: cancellationToken,
+                            continueOnCapturedContext: false)
+                );
+
+                if (enableReporting)
+                {
+                    reportingCancelSrc.Cancel();
+                    await reportingTask;
+                }
+            }
+
+            return (fileNodes, rootNode.TransitiveContentBytes);
+        }
+
+        private static Task StartReportingTask(Action<string> traceOutput, long totalBytes, IDedupUploadSession uploadSession, CancellationTokenSource reportingCancel)
+        {
+            return Task.Run(async () =>
+            {
+                try
+                {
+                    while (!reportingCancel.IsCancellationRequested)
+                    {
+                        traceOutput($"Uploaded {uploadSession.UploadStatistics.TotalContentBytes:N0} out of {totalBytes:N0} bytes.");
+                        await Task.Delay(10000, reportingCancel.Token);
+                    }
+                }
+                catch (OperationCanceledException oce) when (oce.CancellationToken == reportingCancel.Token)
+                {
+                    // Expected
+                }
+                // Print final result
+                traceOutput($"Uploaded {uploadSession.UploadStatistics.TotalContentBytes:N0} out of {totalBytes:N0} bytes.");
+            });
+        }
+
+        private static async Task<List<BlobFileInfo>> GenerateHashes(IReadOnlyList<string> filePaths, CancellationToken cancellationToken)
```

Comment:
You could use PrecomputedHashGenerator here, if the end goal is to get the hash for each file, given the file path - https://dev.azure.com/mseng/AzureDevOps/_git/AzureDevOps?path=%2FArtifactServices%2FShared%2FBlobStore.Common%2FPrecomputedHashesGenerator.cs&_a=contents&version=GBmaster

Reply:
It looks like that class writes the hashes out to a file. I'm not sure if I want that. Is there a way to use it without that?

Reply:
I don't think so. I would say just keep it as-is, since your implementation is fairly straight forward and we can refactor later if need be.